ext/pg_text_decoder.c

```diff
@@ -392,6 +396,204 @@ pg_text_dec_from_base64(t_pg_coder *conv, char *val, int len, int tuple, int fie
 	return out_value;
 }
 
+static inline int char_to_digit(char c)
+{
+  return c - '0';
+}
+
+static int str4_to_int(const char *str)
+{
+  return char_to_digit(str[0]) * 1000
+       + char_to_digit(str[1]) * 100
+       + char_to_digit(str[2]) * 10
+       + char_to_digit(str[3]);
+}
+
+static int str2_to_int(const char *str)
+{
+  return char_to_digit(str[0]) * 10
+       + char_to_digit(str[1]);
+}
+
+static VALUE pg_text_decoder_timestamp_do(t_pg_coder *conv, char *val, int len, int tuple, int field, int enc_idx, int with_timezone)
+{
+  const char *str = val;
+
+  if (isdigit(str[0]) && isdigit(str[1]) && isdigit(str[2]) && isdigit(str[3])
+   && str[4] == '-'
+   && isdigit(str[5]) && isdigit(str[6])
+   && str[7] == '-'
+   && isdigit(str[8]) && isdigit(str[9])
+   && str[10] == ' '
+   && isdigit(str[11]) && isdigit(str[12])
+   && str[13] == ':'
+   && isdigit(str[14]) && isdigit(str[15])
+   && str[16] == ':'
+   && isdigit(str[17]) && isdigit(str[18])
+  )
+  {
+    int year, mon, day;
+    int hour, min, sec;
+    int nsec = 0;
+    int tz_neg = 0;
+    int tz_hour = 0;
+    int tz_min = 0;
+    int tz_sec = 0;
+ 
+    year = str4_to_int(&str[0]);
+    mon = str2_to_int(&str[5]);
+    day = str2_to_int(&str[8]);
+    hour = str2_to_int(&str[11]);
+    min = str2_to_int(&str[14]);
+    sec = str2_to_int(&str[17]);
+    str += 19;
+ 
+    if (str[0] == '.' && isdigit(str[1]))
+    {
+      /* nano second part, up to 9 digits */
+      static const int coef[9] = {
+        100000000, 10000000, 1000000,
+        100000, 10000, 1000, 100, 10, 1
+      };
+      int i;
+
+      str++;
+      for (i = 0; i < 9 && isdigit(*str); i++)
+      {
+        nsec += coef[i] * char_to_digit(*str++);
+      }
+    }
+
+    if (with_timezone)
+    {
+      if ((str[0] == '+' || str[0] == '-') && isdigit(str[1]) && isdigit(str[2]))
+      {
+        tz_neg = str[0] == '-';
+        tz_hour = str2_to_int(&str[1]);
+        str += 3;
+      }
+      if (str[0] == ':')
+      {
+        str++;
+      }
+      if (isdigit(str[0]) && isdigit(str[1]))
+      {
+        tz_min = str2_to_int(str);
+        str += 2;
+      }
+      if (str[0] == ':')
+      {
+        str++;
+      }
+      if (isdigit(str[0]) && isdigit(str[1]))
+      {
+        tz_sec = str2_to_int(str);
+        str += 2;
+      }
+    }
+    if (*str == '\0')
+    {
+#if RUBY_API_VERSION_MAJOR > 2 || (RUBY_API_VERSION_MAJOR == 2 && RUBY_API_VERSION_MINOR >= 3)
+      // must have consumed all the string
+      struct tm tm;
+      tm.tm_year = year - 1900;
```

Comment:
What about timestamps before 1900?

Reply:
In C, the tm_tm_year field stores the number of years since 1900, but this number can be negative.
I added a test that check that date before 1823 and after 2116 are supported